rust/kcl-lib/src/execution/memory.rs

```diff
@@ -156,92 +255,306 @@ pub(crate) const RETURN_NAME: &str = "__return";
+    id: usize,
+    /// Invariant: current_env.1.is_none()
+    current_env: EnvironmentRef,
+    /// Invariant: forall er in call_stack: er.1.is_none()
+    call_stack: Vec<EnvironmentRef>,
 }
 
 // Intended for debugging. Do not rely on this output in any way!
 impl fmt::Display for ProgramMemory {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         let envs: Vec<String> = self
-            .environments
+            .envs()
             .iter()
             .enumerate()
             .map(|(i, env)| format!("{i}: {env}"))
             .collect();
+        write!(
+            f,
+            "ProgramMemory (next stack: {})\nenvs:\n{}",
+            self.next_stack_id.load(Ordering::Relaxed),
+            envs.join("\n")
+        )
+    }
+}
+
+// Intended for debugging. Do not rely on this output in any way!
+impl fmt::Display for Stack {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         let stack: Vec<String> = self
             .call_stack
             .iter()
             .chain(Some(&self.current_env))
             .map(|e| format!("EnvRef({}, {})", e.0, e.1 .0))
             .collect();
-        write!(
-            f,
-            "ProgramMemory\nenvs:\n{}\nstack:\n  {}",
-            envs.join("\n"),
-            stack.join("\n  ")
-        )
+        write!(f, "Stack {}\nstack frames:\n{}", self.id, stack.join("\n"))
     }
 }
 
 impl ProgramMemory {
     #[allow(clippy::new_without_default)]
-    pub fn new() -> Self {
-        Self {
-            environments: vec![],
-            current_env: EnvironmentRef::dummy(),
+    pub fn new() -> Arc<Self> {
+        Arc::new(Self {
+            // Massively over-allocate here to try and avoid reallocating later.
+            environments: UnsafeCell::new(Vec::with_capacity(512)),
             std: None,
-            call_stack: Vec::new(),
             stats: MemoryStats::default(),
+            next_stack_id: AtomicUsize::new(1),
+            write_lock: AtomicBool::new(false),
+        })
+    }
+
+    /// Clone this ProgramMemory.
+    ///
+    /// This is deliberately not a `Clone` impl or called just `clone` since it requires the write
+    /// lock on the memory and so as to be totally unambiguous with cloning an `Arc` of the memory
+    /// (which you should usually prefer).
+    ///
+    /// This is a long-running operation and holds the write lock, which is bad. Callers must ensure
+    /// that no other task will need to use `self` while this runs.
+    fn deep_clone(&self) -> Self {
+        self.with_envs(|envs| Self {
+            environments: UnsafeCell::new(envs.clone()),
+            std: self.std,
+            stats: MemoryStats::default(),
+            next_stack_id: AtomicUsize::new(self.next_stack_id.load(Ordering::Relaxed)),
+            write_lock: AtomicBool::new(false),
+        })
+    }
+
+    pub fn new_stack(self: Arc<Self>) -> Stack {
+        let id = self.next_stack_id.fetch_add(1, Ordering::Relaxed);
```

Comment:
Should this be `Ordering::AcqRel` so that another thread doesn't get the same ID?

Reply:
No, even with relaxed ordering the fetch and add are guaranteed to be atomic, a stronger ordering would only make more guarantees about the order in which the threads got the results, which we don't care about.